R/survival_qp.R

```diff
@@ -0,0 +1,219 @@
+    
+  # Setup the components of the QP and solve
+  if(verbose) message("Creating linear term vector...")
+  q <- create_q_vector_surv(B_X, Bbar_X)
+  
+  if(verbose) message("Creating quadratic term matrix...")
+  P <- create_P_matrix_surv(B_X)
+  
+  I0 <- create_I0_matrix_surv(n)
+  
+  P <- P + lambda * I0
+  
+  if(verbose) message("Creating constraint matrix...")
+  constraints <- create_constraints(n, trt, times, events, t, lowlim, uplim, verbose)
+  
+  settings <- do.call(osqp::osqpSettings,
+                      c(list(verbose = verbose,
+                             eps_rel = eps_rel,
+                             eps_abs = eps_abs),
+                        list(...)))
+  
+  solution <- osqp::solve_osqp(P, q, constraints$A,
+                               constraints$l, constraints$u,
+                               settings)
+  
+  weights <- solution$x
+  
+  # imbalance_root <- t(w) %*% B - Bbar_X
+  # imbalance <- imbalance_root %*% t(imbalance_root)
+  
+  imbalance <- sum(((t(weights) %*% B_X) - Bbar_X)^2) 
+    
+  return(list(weights = weights,
+              imbalance = imbalance))
+}
+
+#' Create the q vector for an QP that solves min_x 0.5 * x'Px + q'x
+#' @param B_X n x k basis matrix of covariates 
+#' @param Bbar_X Vector of population means to re-weight to
+#'
+#' @return q vector
+create_q_vector_surv <- function(B_X, Bbar_X) {
+  
+  q <- -1 * Bbar_X %*% t(B_X)
+  return(q)
+}
+
+#' Create the P matrix for an QP that solves min_x 0.5 * x'Px + q'x
+#' @param B_X n x k basis matrix of covariates 
+#'
+#' @return P matrix
+create_P_matrix_surv <- function(B_X) {
+  
+  P <- B_X %*% t(B_X)
+  return(P)
+}
+
+#' Create diagonal regularization matrix
+#' @param n Total number of units
+#' @param lambda Regularization hyperparameter, default 0
+#' 
+#' @return I0 matrix
+create_I0_matrix_surv <- function(n, lambda) {
+  
+  I0 <- diag(n)
+  return(I0)
+}
+
+#' Create the constraints for QP: l <= Ax <= u
+#' @param n Number of individuals
+#' @param trt Vector of treatment assignments
+#' @param times Vector of event/censoring times (see events for if event or censoring time)
+#' @param events Vector of boolean censoring indicators (whether individual was censored)
+#' @param t Time
+#' @param lowlim Lower limit on weights, default 1
+#' @param uplim Upper limit on weights, default NULL
+#' @param verbose Whether to show messages, default T
+#' 
+#' @return A, l, and u
+create_constraints <- function(n, trt, times, events, t, lowlim = 1, uplim = NULL, verbose = TRUE) {
+
+  # Vector of booleans indicating if indiv is not censored at time t
+  noncens_t <- ((times >= t) & (events == TRUE)) | (events == FALSE)
+
+  if(verbose) message("\tx Summed weight constraint for non-censored treatment group")
+  A1 <- as.integer(noncens_t & (trt == 1))
+  l1 <- n
+  u1 <- n
+  
+  if(verbose) message("\tx Summed weight constraint for non-censored control group")
+  A2 <- as.integer(noncens_t & (trt == 0))
+  l2 <- n
+  u2 <- n
```

Comment:
I think these should solve separate optimization problems for the treated and control units. You can write one version of the code e.g. for only treated units, then pass in `1-trt` to have it work for the control units. Right now I think this mixes the two and doesn't get balance on either.